### ISSUE TITLE
RNN called with seq_lengths gives a ValueError

### ISSUE THREAD

**Issue by giovannic** [google/flax#3208]
### System information
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Mac OS 11.7
- Flax, jax, jaxlib versions (obtain with `pip show flax jax jaxlib`:
jax==0.4.13
jaxlib==0.4.13
flax==0.7.0 (installed from github@main=881c4495a2423d5014ac0175dd490dfa2b6d5ab6)
- Python version:
Python 3.9.13

### Problem you have encountered:

Calling an RNN with a seq_lengths argument gives a ValueError

### What you expected to happen:

The RNN performs a forward pass

### Logs, error messages, etc:

```
...
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../venv/lib/python3.9/site-packages/flax/linen/recurrent.py", line 813, in __call__
    scan_output = scan(self.cell, carry, inputs)
  File ".../venv/lib/python3.9/site-packages/flax/core/axes_scan.py", line 139, in scan_fn
    _, out_pvals, _ = pe.trace_to_jaxpr_nounits(f_flat, in_pvals)
  File ".../venv/lib/python3.9/site-packages/flax/core/axes_scan.py", line 118, in body_fn
    ys = jax.tree_util.tree_map(
ValueError: Expected tuple, got Traced<ShapedArray(float32[2,265]):JaxprTrace(level=1/0)>.
```
### Steps to reproduce:

```python
import jax
import jax.numpy as jnp
import flax.linen as nn

x = jnp.ones((2, 10, 6))
lstm = nn.RNN(nn.LSTMCell(265))
variables = lstm.init(jax.random.PRNGKey(0), x)
y = lstm.apply(variables, x, seq_lengths=jnp.array([5, 5]))
```

**Comment by giovannic** [comment]
Ah, just realised that it works when `return_carry=True`

**Comment by cgarciae** [comment]
Thanks @giovannic for reporting this! There was a bug in the logic, it will be fixed by #3213.
